clang/lib/Sema/SemaExpr.cpp

```diff
@@ -7735,7 +7735,8 @@ ExprResult Sema::BuildResolvedCallExpr(Expr *Fn, NamedDecl *NDecl,
   }
 
   if (CXXMethodDecl *Method = dyn_cast_or_null<CXXMethodDecl>(FDecl))
-    if (Method->isImplicitObjectMemberFunction())
+    if (!isa<RequiresExprBodyDecl>(CurContext) &&
```

Comment:
I don't really understand the change, can you better explain why this is the solution here?

Reply:
In require body, name resolving just lookup function name and don't check function parameter without function call(I think).
```cpp
struct B {
    template <typename S>
    void foo(int);

    void bar();
};

template <typename T, typename S>
struct A : T {
    auto foo() {
        static_assert(requires { T::template foo<S>(); });
        static_assert(requires { T::bar(); });
    }
};

int main() {
    A<B, double> a;
    //a.foo(0);
}
```
After add a parameter to `foo`, this code also compiles correctly without calling.

Reply:
I guess I don't see how 'implicit object member function' is what we want there?  It seems what we really want is 'in the context of a requires clause', right?

Reply:
@erichkeane I am still feeling that we have no need to check whether the method has a explicit object parameter in require clause. In type requirement, checking nested member is enough(name resolution and parameter checking has already completed) and we don't need object argument since it isn't a function call.
```cpp
struct S {
  void foo() {}
};
void bar() {
  S::foo(); // need object parameter
  foo();    // name resolution failed
```
It is only doing a function call that object parameter is required.

Reply:
I still dont think 'implictObjectMemberFunction' is the right thing here.  Why does the fact that it is not a 'Explit Object Member Function' matter?  

like:
```
struct S {
void foo(this S){}
};
void bar() {
S::foo();
};
```
??

Or static member functions?  This solution just doesn't seem right to me, and your responses dont' make it more clear.

Reply:
Do you mean that checking whether `Method` is an implicit object function is redundant? After I removed https://github.com/llvm/llvm-project/blob/844b532713986999aa1ffed0883eff2d1339ec7a/clang/lib/Sema/SemaExpr.cpp#L7723-L7726
These testcase failed

https://github.com/llvm/llvm-project/blob/844b532713986999aa1ffed0883eff2d1339ec7a/clang/test/CXX/drs/dr3xx.cpp#L1084-L1095
https://github.com/llvm/llvm-project/blob/844b532713986999aa1ffed0883eff2d1339ec7a/clang/test/SemaTemplate/instantiate-using-decl.cpp#L150-L168
with no diagnose. Or the checking shouldn't be placed at current position?

Reply:
And if only checking whether `Method` is a static method, we failed on:
https://github.com/llvm/llvm-project/blob/ef9446bd2d362ec90cd681ae59463d16bf671fe8/clang/test/CXX/drs/dr26xx.cpp#L225-L240